### PR TITLE
JAMES-4811 implement JMAP filtering: combine rule conditions

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 1]
```diff
@@ -215,22 +217,45 @@ public static ImmutableList<RuleDTO> from(List<Rule> rules) {
...
+                   @JsonProperty("name") String name,
+                   @JsonProperty("conditions") List<ConditionDTO> conditionDTOs,
+                   @JsonProperty("conditionCombiner") Rule.ConditionCombiner conditionCombiner,
+                   @JsonProperty("action") ActionDTO actionDTO) {
+        this.name = name;
+        this.conditionDTO = null;
+        this.conditionDTOs = conditionDTOs;
+        this.conditionCombiner = conditionCombiner;
+        this.actionDTO = actionDTO;
+        Preconditions.checkNotNull(id);
```
Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:243]: preconditions should go first

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 2]
```diff
@@ -344,15 +350,22 @@ public static Builder builder() {
...
+
     private final Id id;
     private final String name;
-    private final Condition condition;
+    private final List<Condition> conditions;
+    private final ConditionCombiner conditionCombiner;
     private final Action action;
 
-    private Rule(Id id, String name, Condition condition, Action action) {
+    private Rule(Id id, String name, List<Condition> conditions, ConditionCombiner conditionCombiner, Action action) {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java:364]: I would wrap conditions and combiner into its own distinct POJO:

```
class Conditions {
    ConditionCombiner conditionCombiner;
    List<Condition> conditions;
}
```

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java [hunk 3]
```diff
@@ -30,18 +33,51 @@
...
+                    return mailMatchingCondition.getContentMatcher().match(headerLines, mailMatchingCondition.getRuleValue());
+                };
+
+                switch (conditionCombiner) {
+                    case AND:
+                        return mailMatchingConditions.stream().allMatch(predicate);
+                    case OR:
+                        return mailMatchingConditions.stream().anyMatch(predicate);
+                    default:
+                        return false;
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java:66]: Explicitly throw on unmanaged cases?

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 4]
```diff
@@ -725,19 +726,19 @@ void mailDirectiveShouldSetLastMatchedRuleWhenMultipleRules(JMAPFilteringTestSys
                 Rule.builder()
                     .id(Rule.Id.of("1"))
                     .name("rule 1")
-                    .condition(Rule.Condition.of(SUBJECT, CONTAINS, UNSCRAMBLED_SUBJECT))
+                    .conditions(Arrays.asList(Rule.Condition.of(SUBJECT, CONTAINS, UNSCRAMBLED_SUBJECT)))
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:729]: We can keep a convenience method to build a rule with a single condition and avoid boiler plate.

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringExtension.java [hunk 5]
```diff
@@ -112,7 +112,7 @@ public void defineRulesForRecipient1(List<Rule.Condition> conditions) {
                 .map(condition -> Rule.builder()
                     .id(Rule.Id.of(String.valueOf(counter.incrementAndGet())))
                     .name(String.valueOf(counter.incrementAndGet()))
-                    .condition(condition)
+                    .conditions(Arrays.asList(condition))
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringExtension.java:115]: `Arrays.asList` => `ImmutableList.of()` 

Guava is your friend, immutable collections is the good.

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 6]
```diff
@@ -52,6 +52,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
```
Comment (chibenwa) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:52]: No test with several conditions in here?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java (hunk 1 above)
Comment (Arsnael) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:243]: Please refactor the initial constructor as well that has preconditions at the end by mistake

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 7]
```diff
@@ -254,13 +279,21 @@ public ActionDTO getAction() {
...
+            .action(actionDTO.toAction());
+
+        if (conditionDTO != null) {
+            ruleBuilder.conditions(Arrays.asList(conditionDTO.toCondition()));
+        } else {
+            ruleBuilder.conditions(conditionDTOs.stream().map(ConditionDTO::toCondition).collect(Collectors.toList()));
+        }
+
+        if (conditionCombiner != null) ruleBuilder.conditionCombiner(conditionCombiner);
+        else ruleBuilder.conditionCombiner(Rule.ConditionCombiner.AND);
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:294]: No if else blocks like this, should be:
```
if (condition) {
    code
} else {
    code
}
```
better readability

Comment (Arsnael) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:294]: I think we could have kept the initial layout? Could extract the if else condition to functions

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java [hunk 8]
```diff
@@ -51,27 +87,28 @@ private HeaderMatcher(ContentMatcher contentMatcher, String ruleValue,
...
-        Optional<ContentMatcher> maybeContentMatcher = ContentMatcher.asContentMatcher(ruleCondition.getField(), ruleCondition.getComparator());
-        Optional<HeaderExtractor> maybeHeaderExtractor = HeaderExtractor.asHeaderExtractor(ruleCondition.getField());
-
-        return new HeaderMatcher(
-            maybeContentMatcher.orElseThrow(() -> new RuntimeException("No content matcher associated with field " + ruleCondition.getField())),
-            rule.getCondition().getValue(),
-            maybeHeaderExtractor.orElseThrow(() -> new RuntimeException("No content matcher associated with comparator " + ruleCondition.getComparator())));
+        return new HeaderMatcher(rule.getConditions().stream()
+            .map(ruleCondition -> new MailMatchingCondition(
+                ContentMatcher.asContentMatcher(ruleCondition.getField(), ruleCondition.getComparator()).
```
Comment (Arsnael) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java:106]: the `.` at the end should be on the below line with `orEleseThrow`

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 9]
```diff
@@ -217,30 +254,27 @@ public static ImmutableList<RuleDTO> from(List<Rule> rules) {
...
-    private final Rule.ConditionCombiner conditionCombiner;
+    private final ConditionGroupDTO conditionGroupDTO;
     private final ActionDTO actionDTO;
 
     @JsonCreator
     public RuleDTO(@JsonProperty("id") String id,
                    @JsonProperty("name") String name,
-                   @JsonProperty("conditions") List<ConditionDTO> conditionDTOs,
-                   @JsonProperty("conditionCombiner") Rule.ConditionCombiner conditionCombiner,
+                   @JsonProperty("conditionGroup") ConditionGroupDTO conditionGroupDTO,
```
Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java]: I do not think this preserved retro-compatibility...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 10]
```diff
@@ -72,6 +72,54 @@ public String toString() {
...
+            return Objects.hash(conditionCombiner, conditions);
+        }
+
+        @Override
+        public String toString() {
+            return "ConditionCombination{" +
+                "conditionCombiner=" + conditionCombiner +
+                ", conditions=" + conditions +
+                '}';
+        }
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java:120]: Use MoreObjects.toStringHelper

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 11]
```diff
@@ -72,6 +72,54 @@ public String toString() {
...
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ConditionGroup that = (ConditionGroup) o;
+            return conditionCombiner == that.conditionCombiner && Objects.equals(conditions, that.conditions);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(conditionCombiner, conditions);
+        }
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java:112]: Equals Verfier + standard pattern.

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 12]
```diff
@@ -249,15 +283,18 @@ public RuleDTO(@JsonProperty("id") String id,
...
                    @JsonProperty("action") ActionDTO actionDTO) {
+        Preconditions.checkNotNull(id);
+
         this.name = name;
         this.conditionDTO = conditionDTO;
-        this.conditionDTOs = conditionDTOs;
-        this.conditionCombiner = conditionCombiner;
+        if (conditionGroupDTO != null) {
```
Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:295]: Ban null. Use optionals instead. 

Register jackson JDK 8 module if needed.

**Review by chibenwa — COMMENTED** [review]
This is going in the good direction.

I'm expecting that you get quickly used to the project code style and habbits.

I miss tests for both event formats for Cassandra storage CF https://github.com/apache/james-project/tree/master/server/data/data-jmap-cassandra/src/test/resources/json and https://github.com/apache/james-project/blob/master/server/data/data-jmap-cassandra/src/test/java/org/apache/james/jmap/cassandra/filtering/DTOTest.java

(I expect to add at least 3 json files, the v 3, and I expect the assertions to break as well, as the format was updated...)

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 13]
```diff
@@ -254,13 +316,13 @@ public ActionDTO getAction() {
...
             .id(Rule.Id.of(id))
             .name(name)
-            .condition(conditionDTO.toCondition())
-            .name(name)
-            .action(actionDTO.toAction())
-            .build();
+            .conditionGroup(conditionGroupDTO.toConditionGroup())
+            .action(actionDTO.toAction());
+
+        return ruleBuilder.build();
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:325]: No need to create a temporary Rule.Builder object, you could return the all struct directly?

**Comment by hungphan227** [comment]
@chibenwa I have just updated DTOTest. There is are still things that need to be done but could you check if it goes in right direction?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java (hunk 12 above)
Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:293]: Use optionals

**Review by chibenwa — COMMENTED** [review]
Looks great!

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 14]
```diff
@@ -34,6 +34,43 @@
...
+            return new ConditionGroupDTO(conditionGroup.getConditionCombiner(), conditionGroup.getConditions().stream().map(ConditionDTO::from).collect(Collectors.toList()));
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ConditionGroupDTO that = (ConditionGroupDTO) o;
+            return conditionCombiner == that.conditionCombiner && Objects.equals(conditionDTOs, that.conditionDTOs);
+        }
```
[the remark this answers is not in the document]
Comment (hungphan227) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:71]: "Use equalsVerfier". I do not understand what you mean

"Also please follow the equals pattern of other POJOs". Why?

[the remark this answers is not in the document]
Comment (vttranlina) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:71]: > "Use equalsVerfier". I do not understand what you mean

nl.jqno.equalsverifier.EqualsVerifier

ex
```
EqualsVerifier.forClass(MDN.class)
            .verify();
```

[the remark this answers is not in the document]
Comment (hungphan227) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:71]: it seems that EqualsVerifier is only be used for testing

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 15]
```diff
@@ -32,6 +34,50 @@
 
 public class RuleDTO {
 
+    public static class ConditionGroupDTO {
+
+        private Rule.ConditionCombiner conditionCombiner;
+        private List<ConditionDTO> conditionDTOs;
```
Comment (hungphan227) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:40]: final

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 16]
```diff
@@ -215,24 +261,43 @@ public static ImmutableList<RuleDTO> from(List<Rule> rules) {
...
+        this.conditionGroupDTO = conditionGroupDTO;
+        this.actionDTO = actionDTO;
+
+        this.id = id;
+    }
+
     @JsonCreator
     public RuleDTO(@JsonProperty("id") String id,
                    @JsonProperty("name") String name,
                    @JsonProperty("condition") ConditionDTO conditionDTO,
```
Comment (hungphan227) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:289]: Optional also?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 17]
```diff
@@ -215,24 +261,43 @@ public static ImmutableList<RuleDTO> from(List<Rule> rules) {
...
+                   @JsonProperty("conditionGroup") Optional<ConditionGroupDTO> conditionGroupDTO,
                    @JsonProperty("action") ActionDTO actionDTO) {
+        Preconditions.checkNotNull(id);
+
         this.name = name;
-        this.conditionDTO = conditionDTO;
+        if (!conditionGroupDTO.isEmpty()) {
+            this.conditionGroupDTO = conditionGroupDTO.get();
```
Comment (hungphan227) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java:296]: .get is cheating

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 18]
```diff
@@ -72,6 +72,60 @@ public String toString() {
...
+            return new ConditionGroup(conditionCombiner, conditions);
+        }
+
+        public static ConditionGroup of(ConditionCombiner conditionCombiner, Condition... conditions) {
+            return new ConditionGroup(conditionCombiner, ImmutableList.copyOf(conditions));
+        }
+
+        public static ConditionGroup of(Condition condition) {
+            return ConditionGroup.of(ConditionCombiner.AND, condition);
+        }
```
Comment (hungphan227) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java:90]: static methods goes first

File: server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java [hunk 19]
```diff
@@ -51,27 +83,28 @@ private HeaderMatcher(ContentMatcher contentMatcher, String ruleValue,
...
-            rule.getCondition().getValue(),
-            maybeHeaderExtractor.orElseThrow(() -> new RuntimeException("No content matcher associated with comparator " + ruleCondition.getComparator())));
+        return new HeaderMatcher(rule.getConditionGroup().getConditions().stream()
+            .map(ruleCondition -> new MailMatchingCondition(
+                ContentMatcher.asContentMatcher(ruleCondition.getField(), ruleCondition.getComparator())
+                    .orElseThrow(() -> new RuntimeException("No content matcher associated with field " + ruleCondition.getField())),
+                ruleCondition.getValue(),
+                HeaderExtractor.asHeaderExtractor(ruleCondition.getField())
+                    .orElseThrow(() -> new RuntimeException("No content matcher associated with comparator " + ruleCondition.getComparator())))
+            ).collect(Collectors.toList()), rule.getConditionGroup().getConditionCombiner());
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/mailet/filter/MailMatcher.java:107]: `Collectors.toList()` => `ImmutableList.toImmutableList()`

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 20]
```diff
@@ -725,26 +727,26 @@ void mailDirectiveShouldSetLastMatchedRuleWhenMultipleRules(JMAPFilteringTestSys
                 Rule.builder()
                     .id(Rule.Id.of("1"))
                     .name("rule 1")
-                    .condition(Rule.Condition.of(SUBJECT, CONTAINS, UNSCRAMBLED_SUBJECT))
+                    .conditionGroup(Rule.Condition.of(SUBJECT, CONTAINS, UNSCRAMBLED_SUBJECT))
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:730]: Uneeded change?

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 21]
```diff
@@ -1057,4 +1059,79 @@ void actionShouldCombineFlags(JMAPFilteringTestSystem testSystem) throws Excepti
...
+    @Test
+    void shouldMatchRuleOfAndConditionCombiner(JMAPFilteringTestSystem testSystem) throws Exception {
+        Mono.from(testSystem.getFilteringManagement().defineRulesForUser(RECIPIENT_1_USERNAME,
+            Optional.empty(),
+            Rule.builder()
+                .id(Rule.Id.of("1"))
+                .name("rule 1")
+                .conditionGroup(Rule.ConditionGroup.of(Rule.ConditionCombiner.AND, Rule.Condition.of(FROM, CONTAINS, USER_2_USERNAME), Rule.Condition.of(SUBJECT, CONTAINS, "cd")))
+                .action(Rule.Action.of(Rule.Action.AppendInMailboxes.withMailboxIds(testSystem.getRecipient1MailboxId().serialize()),
+                    false, false, false, ImmutableList.of()))
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:1072]: Use a constant in RuleFixture for the action

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 22]
```diff
@@ -1057,4 +1059,79 @@ void actionShouldCombineFlags(JMAPFilteringTestSystem testSystem) throws Excepti
         assertThat(StorageDirective.fromMail(Username.of("recipient1"), mail).getFlags().get())
             .isEqualTo(expectedFlags);
     }
+
+    @Test
+    void shouldMatchRuleOfAndConditionCombiner(JMAPFilteringTestSystem testSystem) throws Exception {
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:1064]: andShouldMatchWhenAllConditionAreMet

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 23]
```diff
@@ -1057,4 +1059,79 @@ void actionShouldCombineFlags(JMAPFilteringTestSystem testSystem) throws Excepti
...
+        FakeMail mail = testSystem.asMail(mimeMessageBuilder().addHeader(FROM.asString(), USER_2_ADDRESS).addHeader(SUBJECT.asString(), "abcdef"));
+
+        testSystem.getJmapFiltering().service(mail);
+
+        assertThatAttribute(mail.getAttribute(RECIPIENT_1_USERNAME_ATTRIBUTE_NAME))
+            .isEqualTo(RECIPIENT_1_MAILBOX_1_ATTRIBUTE);
+    }
+
+    @Test
+    void shouldNotMatchRuleOfAndConditionCombiner(JMAPFilteringTestSystem testSystem) throws Exception {
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:1084]: andShouldNotMatchWhenSomeConditionsAreMet

Test case naming structure:

XXXXX Should YYYYY When ZZZZ

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 24]
```diff
@@ -1057,4 +1059,79 @@ void actionShouldCombineFlags(JMAPFilteringTestSystem testSystem) throws Excepti
...
+                .build())).block();
+
+        FakeMail mail = testSystem.asMail(mimeMessageBuilder().addHeader(FROM.asString(), USER_2_ADDRESS).addHeader(SUBJECT.asString(), "abcdef"));
+        testSystem.getJmapFiltering().service(mail);
+
+        assertThat(mail.getAttribute(RECIPIENT_1_USERNAME_ATTRIBUTE_NAME).isEmpty()).isTrue();
+    }
+
+    @Test
+    void shouldMatchRuleOfOrConditionCombiner(JMAPFilteringTestSystem testSystem) throws Exception {
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:1102]: orShouldMatchWhenAllConditionsAreMet

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 25]
```diff
@@ -1057,4 +1059,79 @@ void actionShouldCombineFlags(JMAPFilteringTestSystem testSystem) throws Excepti
...
+
+        FakeMail mail = testSystem.asMail(mimeMessageBuilder().addHeader(FROM.asString(), USER_2_ADDRESS).addHeader(SUBJECT.asString(), "abcdef"));
+        testSystem.getJmapFiltering().service(mail);
+
+        assertThatAttribute(mail.getAttribute(RECIPIENT_1_USERNAME_ATTRIBUTE_NAME))
+            .isEqualTo(RECIPIENT_1_MAILBOX_1_ATTRIBUTE);
+    }
+
+    @Test
+    void shouldNotMatchRuleOfOrConditionCombiner(JMAPFilteringTestSystem testSystem) throws Exception {
```
Comment (hungphan227) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:1121]: orShouldNotMatchWhenNoConditionsAreMet

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java (hunk 21 above)
Comment (hungphan227) [reply server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java]: Could not find the same thing in RuleFixture. Moreover, RuleFixture is in a different module.

**Comment by chibenwa** [comment]
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:check (check-style) on project james-server-jmap-draft: You have 1 Checkstyle violation. -> [Help 1]`

Please compile locally.

File: server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java [hunk 26]
```diff
@@ -1057,4 +1058,75 @@ void actionShouldCombineFlags(JMAPFilteringTestSystem testSystem) throws Excepti
...
+        FakeMail mail = testSystem.asMail(mimeMessageBuilder().addHeader(FROM.asString(), USER_2_ADDRESS).addHeader(SUBJECT.asString(), "abcdef"));
+
+        testSystem.getJmapFiltering().service(mail);
+
+        assertThatAttribute(mail.getAttribute(RECIPIENT_1_USERNAME_ATTRIBUTE_NAME))
+            .isEqualTo(RECIPIENT_1_MAILBOX_1_ATTRIBUTE);
+    }
+
+    @Test
+    void andShouldNotMatchWhenSomeConditionsAreMet(JMAPFilteringTestSystem testSystem) throws Exception {
```
Comment (Arsnael) [inline server/protocols/jmap-draft/src/test/java/org/apache/james/jmap/mailet/filter/JMAPFilteringTest.java:1082]:
```suggestion
    void andShouldNotMatchWhenSomeConditionsAreNotMet(JMAPFilteringTestSystem testSystem) throws Exception {
```

**Review by Arsnael — COMMENTED** [review]
Just a typo, overall it looks good to me, good job :)

**Review by Arsnael — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Pull request merged**